app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -535,4 +465,156 @@ public int setSelectionExpandWholeLines() {
+            if (!_editor.getLocalVisibleRect(_visibleRect)) {
+                return;
+            }
+
+            final CharSequence text = _editor.getText();
+            final Layout layout = _editor.getLayout();
+            if (text == null || layout == null) {
+                return;
+            }
+
+            final int offsetY = _editor.getPaddingTop();
+            final int layoutLineCount = layout.getLineCount();
+
+            // Only if the visible area or the layout line count changed
+            // We make a single pass through the text to determine
+            // 1. y positions and line numbers we want to draw
+            // 2. max line number (to gauge gutter width)
+            if (isVisibleAreaChanged() || layoutLineCount != _lastLayoutLineCount) {
+                _lastLayoutLineCount = layoutLineCount;
+                _maxNumber = 1;
+                _startNumber = 1;
+                _numberYPositions.clear();
+                _top = _visibleRect.top - _visibleRect.height();
+                _bottom = _visibleRect.bottom + _visibleRect.height();
+
+                for (int i = 0; i < layoutLineCount; i++) {
+                    System.err.println(i);
+                    final int start = layout.getLineStart(i);
+                    if (start == 0 || text.charAt(start - 1) == '\n') {
+                        final int y = layout.getLineBaseline(i);
+                        if (y < _top) {
+                            _startNumber++;
+                        } else if (y < _bottom) {
+                            _numberYPositions.add(y);
+                        }
+                        _maxNumber++;
+                    }
+                }
+            }
+
+            // Only if the text size or the max line number of digits changed
+            // Update the gutter parameters and set padding left
+            if (_paint.getTextSize() != _lastTextSize || getNumberDigits(_maxNumber) != _lastMaxNumberDigits) {
+                final int width = Math.round(_paint.measureText(String.valueOf(_maxNumber)));
+                _numberX = LINE_NUMBERS_PADDING_LEFT + width;
+                _gutterLineX = _numberX + LINE_NUMBERS_PADDING_RIGHT;
+                _editor.setPadding(_gutterLineX + 10, _editor.getPaddingTop(), _editor.getPaddingRight(), _editor.getPaddingBottom());
+            }
+
+            // Draw the gutter line
+            _paint.setColor(Color.LTGRAY);
+            canvas.drawLine(_gutterLineX, _top, _gutterLineX, _bottom, _paint);
+
+            // Draw the line numbers
+            _paint.setColor(Color.GRAY);
+            for (int i = 0; i < _numberYPositions.size(); i++) {
+                final String number = String.valueOf(_startNumber + i);
+                canvas.drawText(number, _numberX, _numberYPositions.get(i) + offsetY, _paint);
+            }
+        }
+
+        /**
+         * Suspend drawing the line numbers.
+         */
+        public void suspend() {
```

Comment:
IMO this method should be called `clear` and should be called only once when line numbers are disabled, not every time in onDraw.